File: leetcode/problems_213_house-robber-ii.py
import unittest

class Solution(object):
    def rob1(self, nums):
        if not nums:
            return 0
        elif len(nums) == 1:
            return nums[0]

        # mid >= 1
        mid = len(nums) // 2
        return max(self.rob1(nums[:mid-1]) + nums[mid] + self.rob1(nums[mid+2:]),
                self.rob1(nums[:mid]) + self.rob1(nums[mid+1:]))

    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        if not nums:
            return 0
        elif len(nums) == 1:
            return nums[0]

        return max(self.rob1(nums[1:]), self.rob1(nums[:-1]))
        
class Test(unittest.TestCase):
    def setUp(self):
        self.inst = Solution()

    def test_sample_0(self):
        self.assertEqual(self.inst.rob([1]), 1)

    def test_sample_1(self):
        self.assertEqual(self.inst.rob([1, 3, 1]), 3)

    def test_sample_2(self):
        self.assertEqual(self.inst.rob([1, 3, 3, 1]), 4)

if __name__ == '__main__':
    unittest.main()

File: leetcodecn/problems_213_house-robber-ii.java
class Solution {
    int dp(int[] nums, int left, int right) {
	int pre = 0; // cur - 1
	int cur = 0; // 
	int tmp = 0;
	for (int i = left; i < right;, i += 1) {
	    tmp = cur;
	    cur = Math.max(pre + nums[i], tmp);
	    pre = tmp;
	}

	return cur;
    }

    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        } else if (nums.length == 1) {
            return nums[0];
	}

        return Math.max(dp(nums, 0, nums.length - 1),
			dp(nums, 1, nums.length));
    }
}
